envoy/http/async_client.h

```diff
@@ -163,13 +168,57 @@ class AsyncClient {
      */
     virtual void reset() PURE;
 
+    /***
+     * Register callback to be called on stream destruction. This callback must persist beyond the
+     * lifetime of the stream or be unregistered via removeDestructorCallback. Calling this method
+     * twice with different callbacks will overwrite.
+     */
+    virtual void setDestructorCallback(StreamDestructorCallbacks callback) PURE;
+
+    /***
+     * Remove previously set destructor callback.
+     */
+    virtual void removeDestructorCallback() PURE;
+
+    /***
+     * Register a callback to be called when high/low watermark events occur on the stream. This
```

Comment:
this is for upstream write buffer watermarks yeah?  maybe worth clarifying since watermarks are confusing AF to everyone, self included =P